Bug 1293512 - Support QR Code on iOS

lemonYLX opened the pull request:
update pull request to swift 3

farhanpatel commented:
<!--
  0 Errors
  0 Warnings
  0 Messages
  1 Markdown
-->

### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
QRCodeViewController.swift | 141 | if,for,while,do statements shouldn't wrap their conditionals in parentheses. 
QRCodeViewController.swift | 262 | Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

lemonYLX commented:
update pull request

bhavnapanchal commented on Client/Frontend/Browser/QRCodeViewController.swift:
```diff
@@ -0,0 +1,265 @@
+            self.captureSession.stopRunning()
+            stopScanLineAnimation()
+            let metaData = metadataObjects.first as! AVMetadataMachineReadableCodeObject
+            let url = NSURL(string: metaData.stringValue)
```
Nice piece of code, I was just wonder what if url is invalid
in that case application might crash. 
will you please test it for that and let me know

else all looks nicely structured and maintained, can't wait to see this feature in latest build.

lemonYLX replied:
I want to test this case, and this case can only build on iPhone or iPad, but I found that I was deleted form the apple developer group.
So could you help to add lyang@mozilla.com to the apple developer group and test group?

bhavnapanchal replied:
Well you can test it by replacing  
let url = NSURL(string: metaData.stringValue)
with 
let url = NSURL(string:"someInvalidURLString")

lemonYLX replied:
I think there are some problems.
Firstly, AVCaptureMetadataOutputObjectsDelegate will call when a capture metadata output object receives relevant metadata objects.
hence, we need used AVCaptureDeviceInput to capture data from an AVCaptureDevice object.
AVCaptureDevice.defaultDevice(withMediaType: AVMediaTypeVideo)
But AVMediaTypeVideo can only build on iPhone / iPad.
In brief, I need use physical device to test this case.
If you have any idea, please tell me. Thanks.

lemonYLX commented:
@bhavnapanchal 
I have tested this case.
If url is invalid, browser will open url in new tab, then prompt users that “Firefox cannot open the page because it has an invalid address”

There is another issue I met during the test:
"QR Code" button should displayed when New Tab is a blank page.
What do you think？

bhavnapanchal commented:
@lemonYLX 
Regarding
------------------------
If url is invalid, browser will open url in new tab, then prompt users that “Firefox cannot open the page because it has an invalid address”
------------------------
> So finally you able to build it 👍 .
It sounds good that it is not crashing.
My personal opinion is that it suppose to be perform search operation on google or other search engine when QR code is not valid URL.

OR

In QR code screen it suppose to be pop-up message that this is invalid url you want to search with this text.
Benefit of this approach is that if text contain some adults string it will not open in big screen. 

Regarding
------------------------
There is another issue I met during the test:
"QR Code" button should displayed when New Tab is a blank page.
What do you think？
------------------------
> Well I think this is nice case of user experience where user can read QR code and open web-page. But I am not sure about UI implementation for that how it looks.

OR

We can add it in option menu so that user can open QR code reader whenever she/he wants.
Benefits for above approach is you do not have to compromise stander UI of blank page
and user do not have to open blank page to open QR code reader.

I am also a swift developer and let me know if you need any help in building above things
and let me know your comments too.

lemonYLX commented:
@bhavnapanchal 
Regarding 

> My personal opinion is that it suppose to be perform search operation on google or other search engine when QR code is not valid URL.

I totally agree with you, and I have updated my PR.
If url is invalid, ff will perform search operation on search engine.

And,
I have added "QR Code"button in the menu when New Tab is a blank page.

Please help me review it.  Thank you.

bhavnapanchal commented:
@lemonYLX 

Thanks lemon,
Now code looks fine, But i could not check it with my device as there are some error while running 'sh ./bootstrap.sh'

'Build Failed
	Task failed with exit code 65:
	/usr/bin/xcrun xcodebuild -workspace /Users/jageen.shukla/Documents/JageenGit/firefox-ios/Carthage/Checkouts/Alamofire/Alamofire.xcworkspace -scheme "Alamofire iOS" -configuration Release -derivedDataPath /Users/jageen.shukla/Library/Caches/org.carthage.CarthageKit/DerivedData/Alamofire/2.0.2 -sdk iphoneos ONLY_ACTIVE_ARCH=NO BITCODE_GENERATION_MODE=bitcode CODE_SIGNING_REQUIRED=NO CODE_SIGN_IDENTITY= CARTHAGE=YES clean build'

I review your code and it looks, clean to go just one case I was worried that you stop self.captureSession.stopRunning() in viewWillDisappear but did not start it on viewWillAppear.
I just check code I am not sure about the case. sorry if I pointed out irrelevant issue.

Regards
Bhavna.

lemonYLX commented:
@bhavnapanchal 

> But i could not check it with my device as there are some error while running 'sh ./bootstrap.sh'

I am so sorry to hear that. You can try 
remove any local copies of your Carthage dependencies by removing the *Carthage* folder and
*Cartfile.resolved* file from the project's root directory. Afterwards, re-run the *./bootstrap.sh* command to checkout fresh copies of the Swift 3 dependencies.
But I am not sure that can help you.

Regarding
>  I was worried that you stop self.captureSession.stopRunning() in viewWillDisappear but did not start it on viewWillAppear.

According to my test, there is no another issue. If you have any questions, please contact me. 

Regards
Linxiu Yang

lemonYLX commented:
![](https://github.trello.services/images/mini-trello-icon.png) [QR Code Scanner](https://trello.com/c/LFpiXygY/31-qr-code-scanner)

lemonYLX commented:
![](https://github.trello.services/images/mini-trello-icon.png) [QR Code Scanner](https://trello.com/c/LFpiXygY/31-qr-code-scanner)

farhanpatel commented:
We merged a separate. If this PR is still relevant. Feel free to rebase and reopen.
